### PR TITLE
Add 3DES and file support to des.py

### PR DESCRIPTION
**Pull request by elp2**
python des.py -c 1234567890abcdef  d3001515557aadadd3001515557aadae
python des.py -d 161f79c6910b9c2d d3001515557aadadd3001515557aadae
-> 1234567890abcdef

### CODE REVIEW

**Comment by mbrown1413** [comment]
Sorry for the late response. I'm out of town at the moment, but I will look at this next week.

File: des.py [hunk 1]
```diff
@@ -153,6 +152,51 @@ def bits_to_pretty(bits, blocksize=8):
...
+    bit_rows = []
+    with open(filename, "rb") as f:
+        while True:
+            byte = f.read(8)
+            if byte == "":
+                break
+            bits = bittools.ascii_to_bits(byte)
+            while len(bits) < 64:
+                bits = bits + [0,0,0,0,0,0,0,0]
+            log(bits)
```
Comment (mbrown1413) [inline des.py]: Avoid printing/logging extra information like intermediate results.
```suggestion
```

File: des.py [hunk 2]
```diff
@@ -167,39 +211,69 @@ def bits_to_pretty(bits, blocksize=8):
...
+        else:
+            # 3DES is EDE / DED, so round 2 is opposite.
+            decrypt = not options.decrypt
+
+        result = []
+        for t in text:
+            if decrypt:
+                result += dsa_decrypt(t, key)
+            else:
+                result += dsa_encrypt(t, key)
```
Comment (mbrown1413) [inline des.py:267]: The same value for `text` is used for each round here, but each round should start with the previous round's result.

File: des.py [hunk 3]
```diff
@@ -167,39 +211,69 @@ def bits_to_pretty(bits, blocksize=8):
...
+
+        if options.ascii and decrypt:
+            print bittools.bits_to_ascii(result)
+        else:
+            print bittools.bits_to_hex(result)
```
Comment (mbrown1413) [inline des.py]: You can unindent this whole block of code that prints/outputs the results, so only the result of the final round is shown.

**Review by mbrown1413 — CHANGES_REQUESTED** [review]
Hi @elp2,

Thanks for waiting, and thanks for the new features! I like how it automatically detects the keying method based on the length.

I've made a few comments in a review. There is a small problem with how the results from each round are passed to the next round.

File: des.py (hunk 2 above)
Comment (elp2) [reply des.py:267]: Nice catch!

**Review by elp2 — COMMENTED** [review]
Double checked the encryption / decryption. Example:

python des.py -c 6173636963687273 0123456789abcdef00000000000000001111111111111111
32709d10300c0ed8
python des.py -d 32709d10300c0ed8 0123456789abcdef00000000000000001111111111111111
6173636963687273

File: des.py (hunk 1 above)
Comment (elp2) [reply des.py]: Fixed in round 2

**Review by elp2 — COMMENTED** [review]
Not sure if I resolved the log(bits) properly in #1. Looks good now in #2.

**Comment by mbrown1413** [comment]
Thanks for all this! It all looks good now, including your other PRs. Python 3 has been out for so long it feels weird to see xrange.

**Pull request merged**